src/polymer/polymer2-mixin-scanner.ts

```diff
@@ -0,0 +1,140 @@
+/**
+ * @license
+ * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as estree from 'estree';
+
+import {Visitor} from '../javascript/estree-visitor';
+import * as esutil from '../javascript/esutil';
+import {JavaScriptDocument} from '../javascript/javascript-document';
+import {JavaScriptScanner} from '../javascript/javascript-scanner';
+import * as jsdoc from '../javascript/jsdoc';
+
+import {ScannedPolymerElementMixin} from './polymer-element-mixin';
+import {getConfig, getProperties} from './polymer2-config';
+
+export class Polymer2MixinScanner implements JavaScriptScanner {
+  async scan(
+      document: JavaScriptDocument, visit: (visitor: Visitor) => Promise<void>):
+      Promise<ScannedPolymerElementMixin[]> {
+    let visitor = new MixinVisitor(document);
+    await visit(visitor);
+    return visitor._mixins;
+  }
+}
+
+class MixinVisitor implements Visitor {
+  _mixins: ScannedPolymerElementMixin[] = [];
+  private _document: JavaScriptDocument;
+
+  private _currentMixin: ScannedPolymerElementMixin|null = null;
+  private _currentMixinNode: estree.Node|null = null;
+  private _currentMixinFunction: estree.BaseFunction|null = null;
+
+  constructor(document: JavaScriptDocument) {
+    this._document = document;
+  }
+
+  enterFunctionDeclaration(
+      node: estree.FunctionDeclaration, _parent: estree.Node) {
+    const comment = esutil.getAttachedComment(node) || '';
+    const docs = jsdoc.parseJsdoc(comment);
+    if (this._hasPolymerMixinDocTag(docs)) {
+      const name = node.id.name;
+      this._currentMixinFunction = node;
+      this._currentMixin = new ScannedPolymerElementMixin({
+        tagName: name,
+      });
+      this._currentMixinNode = node;
+      this._currentMixin.tagName = name;
+      this._mixins.push(this._currentMixin);
```

Comment:
Here and elsewhere, we should give the best source range that we can for the mixin. When we add the mixin to the array, it seems like a good time to check to see if it has a source range yet, and if not, give it the best one available.